container/push.bzl

```diff
@@ -29,100 +29,198 @@ load(
 
-    pusher_runfiles = [ctx.executable._pusher] + pusher_input
-    runfiles = ctx.runfiles(files = pusher_runfiles)
-    runfiles = runfiles.merge(ctx.attr._pusher[DefaultInfo].default_runfiles)
+def _pusher_write_script(ctx, image, ii, template_file, output_file):
+    args, inputs, md = _pusher_write_scripts_common(ctx, ii)
+
+    img_args, img_inputs = _gen_img_args(ctx, image, _get_runfile_path)
+    args += img_args
+    inputs += img_inputs
 
-    exe = ctx.actions.declare_file(ctx.label.name + ctx.attr.extension)
     ctx.actions.expand_template(
-        template = ctx.file.tag_tpl,
-        output = exe,
+        template = template_file,
+        output = output_file,
         substitutions = {
-            "%{args}": " ".join(pusher_args),
+            "%{args}": " ".join(args),
             "%{container_pusher}": _get_runfile_path(ctx, ctx.executable._pusher),
         },
         is_executable = True,
     )
 
+    runfiles = ctx.runfiles(files = [ctx.executable._pusher] + inputs)
+    runfiles = runfiles.merge(ctx.attr._pusher[DefaultInfo].default_runfiles)
+
+    return runfiles, md
+
+def _pusher_index_write_script(ctx, images, ii, template_file, output_file):
+    args, inputs, md = _pusher_write_scripts_common(ctx, ii)
+
+    for image_platform, image in images.items():
+        platform_args = ["--os", image_platform.os, "--arch", image_platform.arch]
+        if image_platform.variant != "":
+            platform_args += ["--variant", image_platform.variant]
+
+        img_args, img_inputs = _gen_img_args(ctx, image, _get_runfile_path)
+
+        args += ["--"] + platform_args + img_args
+        inputs += img_inputs
+
+    ctx.actions.expand_template(
+        template = template_file,
+        output = output_file,
+        substitutions = {
+            "%{args}": " ".join(args),
+            "%{container_pusher}": _get_runfile_path(ctx, ctx.executable._pusher_index),
+        },
+        is_executable = True,
+    )
+
+    runfiles = ctx.runfiles(
+        files = [ctx.executable._pusher_index] + inputs,
+    )
+    runfiles = runfiles.merge(ctx.attr._pusher_index[DefaultInfo].default_runfiles)
+
+    return runfiles, md
+
+def _get_image_from_target(ctx, target):
+    image = _get_layers(ctx, ctx.label.name, target)
+    if image.get("legacy"):
+        # buildifier: disable=print
+        print("Pushing an image based on a tarball can be very " +
+              "expensive. If the image set on %s is the output of a " % ctx.label +
+              "docker_build, consider dropping the '.tar' extension. " +
+              "If the image is checked in, consider using " +
+              "container_import instead.")
+
+    return image
+
+def _container_push_impl(ctx):
+    """Core implementation of container_push."""
+
+    # TODO: Possible optimization for efficiently pushing intermediate format after container_image is refactored, similar with the old python implementation, e.g., push-by-layer.
```

Comment:
?

Reply:
It's just a code move. This comment already existed.